### DISCUSSION TITLE
What's the best way to set up persistent port forwarding in MeshCentral

### DISCUSSION THREAD

**Discussion by pptman** [Ylianst/MeshCentral#6327] [Q&A] [answered]
As a simple case I want to forward the ssh port from a remote machine (ubuntu) to my local machine (MacOs) through MeshCentral so that I can log in from a local terminal window.  I can do this using either the `MeshCentral Router` app on my Mac or using the `RoutePlus` plug-in in the web interface.  Both work and I can ssh successfully into my remote linux box, but then I get kicked out a short time later with an event recorded in the web interface with message:
```
Ended relay session 1reFO15cOoIq" from 172.18.0.1 to 172.18.0.1 - 300 second(s)"
```
At the same time ssh drops the connection with a message like:
```
Bad packet length 2389489520.
ssh_dispatch_run_fatal: Connection to 127.0.0.1 port 222: Connection corrupted
```

Typically this timeout is 300 seconds, (and in fact I also see a timeout with normal web based terminal sessions that is equally irritating).  And the timeout happens even if there is activity across the ssh connection.

Actually my real use case is to be able to forward half a dozen ports from the remote machine to the local machine that I communicate over using web sockets from a client on the local machine to several servers on the remote machine.  Clearly the 300 second timeout (that happens in this case too) makes it impossible to do anything over a longer period of time without restarting all the relay sessions.

I am new to using MeshCentral so apologies if this is a trivial question. I have managed to do this using other RMM management software with more persistent ssh port forwarding, but would much rather use an open-source self-hosted solution like MeshCentral.

**Comment by si458** [comment]
you can try meshcmd with tcp port forwarding instead https://ylianst.github.io/MeshCentral/meshcmd/
how are you running meshcentralrouter exactly on osx?
i use it with wine 9.0 and dont have any issues?

**Reply by pptman** [reply]
I got meshcentralrouter from https://meshcentral.com/downloads.html , but I will be the first to admit that it looks quite old (v. 0.0l.4) and appears to drop the connection way before the 300s timeout that the server reports.  Likewise the MeshCentral-RoutePlus plugin is at least 2 years old according to its GitHub.  I'm not sure if the problem lies with these tools though as they do at least set up the route correctly, albethey a bit clunky.  I have not tried meshcmd simply because of the lack of a MacOs version, and many previous unsatisfactory experiences with wine!  I will try your suggestion, and the native binaries on a windows machine to see if there is any difference in behaviour, or whether it really is my MeshCentral Server that is dropping the connection.  It is behind an nginx reverse proxy, by the way, but I have increased the timeouts there to > 300 s already as suggested on the MeshCentral2 install pages.

**Reply by si458** [reply]
Yes don't use the meshcentralrouter osx version as it hasn't been updated in years!
Best try running wine and using the windows exe version from ur meshcentral server as that seems to work a treat here for me!

**Reply by pptman** [reply]
Ok, I can confirm that wine + meshcmd does work on MacOs, but can it route several ports at once?  MeshCentralRouter works on Windows but the MeshCentral-RoutePlus plugin still gives 300 second timeouts even when the endpoint is on a windows machine instead of MacOs.  At least the plugin option is just a few hundred lines of javascript, maybe I will take a look there to see if there is a quick fix.

**Reply by si458** [reply, marked as answer]
you can download meshcmd natively for mac from your meshcentral web ui,
login, at bottom click meshcmd, then change it to macos (arm or x86) and download that version
you dont need wine for that.
plugins are unoffically support so sadly we cannot help you if you have plugins enabled on your server
i have never used MeshCentral-RoutePlus so i dont know what it does
but the best thing to try is using meshcmd natively on your mac

**Reply by pptman** [reply]
So I can!  I was confused as in Safari the download link still gives (and downloads) the Windows .exe file even if you select MacOs in the menu.  In Chrome, the link still says Windows .exe but actually downloads the macOS version once selected in the menu.  The only browser that I have tried and that works properly is MS Edge where selecting a version in the pull down menu does change the link.  The macOS version of meshcmd works fine and will even set up multiple routes if run separately for each route at least.  Can the mesh action.txt file rewritten to open multiple routes at once?

**Reply by si458** [reply]
@pptman OOOO another bug maybe! i will look into it as i didnt actually download to test! i just know the options are there!
as for the multiple routes, im not 100% sure as ive personally never used the meshcmd
i just know that you can use meshcmd to do the port forwarding more permantly if you needed too

**Reply by si458** [reply]
ok can confirm bug with safari/chrome(+edge), no issues with firefox so looking into it now!

**Reply by si458** [reply]
ok fixed the tiny bug for you! https://github.com/Ylianst/MeshCentral/commit/cee181fb614fd8e4b4a365866b2302eba812ece7
simple edit default.handlebars and change `onclick` to `onchange` inside the `p10showMeshCmdDialog` function
(around line 8903)

**Reply by pptman** [reply]
Thanks of the help and finding the bug.  I can set up what I need by using the downloaded meshaction.txt file to define the server and remote target parameters and then individual meshcmd instances with command line parameters to define each port mapping that I want.

**Reply by si458** [reply]
Glad it worked! and thank u for finding the 🐛 in the first place! It's hard to test every browser and OS haha
